File: tests/test_e2e.js
/**
 * E2E Integration Test Suite for li-pdf Platform.
 * Tests:
 * 1. Healthcheck & Capability Registry
 * 2. Presigned Upload Flow
 * 3. Idempotent Job Creation (PNG -> WebP conversion & PDF ops)
 * 4. State Machine (QUEUED -> PROCESSING -> VALIDATING -> COMPLETED)
 * 5. Output Validator Verification & Download
 */

const http = require('http');

const API_BASE = process.env.API_BASE || 'http://127.0.0.1:8085';

function log(msg) {
  console.log(`\x1b[34m[TEST]\x1b[0m ${msg}`);
}

function success(msg) {
  console.log(`\x1b[32m[PASS]\x1b[0m ${msg}`);
}

function fail(msg) {
  console.error(`\x1b[31m[FAIL]\x1b[0m ${msg}`);
  process.exit(1);
}

async function request(url, options = {}, body = null) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const reqOptions = {
      hostname: parsed.hostname,
      port: parsed.port,
      path: parsed.pathname + parsed.search,
      method: options.method || 'GET',
      headers: options.headers || {},
    };

    const req = http.request(reqOptions, (res) => {
      const chunks = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end', () => {
        const buffer = Buffer.concat(chunks);
        const text = buffer.toString('utf-8');
        let json = null;
        try {
          json = JSON.parse(text);
        } catch (_) {}
        resolve({ status: res.statusCode, headers: res.headers, body: json, text, buffer });
      });
    });

    req.on('error', reject);
    if (body) {
      req.write(body);
    }
    req.end();
  });
}

async function runTests() {
  log(`Connecting to li-pdf API at ${API_BASE}...`);

  // 1. Healthcheck
  const health = await request(`${API_BASE}/healthz`);
  if (health.status === 200 && health.body?.status === 'healthy') {
    success('Healthcheck passed');
  } else {
    fail(`Healthcheck failed: ${health.status} ${health.text}`);
  }

  // 2. Capabilities
  const capsRes = await request(`${API_BASE}/v1/capabilities`);
  const caps = capsRes.body?.capabilities || [];
  if (capsRes.status === 200 && caps.length > 0) {
    success(`Capability Registry active (${caps.length} operations registered)`);
  } else {
    fail(`Capabilities fetch failed: ${capsRes.status} ${capsRes.text}`);
  }

  // 3. Create a synthetic test PNG file
  log('Creating synthetic test PNG...');
  const pngBytes = Buffer.from(
    'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==',
    'base64'
  );

  // Request Presigned Upload
  const presignRes = await request(
    `${API_BASE}/v1/uploads/presign`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
    },
    JSON.stringify({
      filename: 'sample_test.png',
      size_bytes: pngBytes.length,
      mime_type: 'image/png',
    })
  );

  if (presignRes.status !== 201 || !presignRes.body?.upload_url) {
    fail(`Presign upload failed: ${presignRes.status} ${presignRes.text}`);
  }

  const uploadId = presignRes.body.upload_id;
  const uploadUrl = presignRes.body.upload_url.replace('http://localhost:8085', API_BASE);
  success(`Presigned upload slot created (ID: ${uploadId})`);

  // Direct Upload to Storage URL
  log(`Executing direct upload (${pngBytes.length} bytes)...`);
  const directUpload = await request(
    uploadUrl,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'image/png' },
    },
    pngBytes
  );

  if (directUpload.status !== 200 && directUpload.status !== 204) {
    fail(`Direct upload to storage failed with status ${directUpload.status}`);
  }
  success('Direct storage payload upload verified');

  // 4. Create Job with Idempotency Key
  const idempotencyKey = `e2e_idemp_${Date.now()}`;
  log(`Creating job: 'png_to_webp' with Idempotency-Key: ${idempotencyKey}...`);

  const jobRes = await request(
    `${API_BASE}/v1/jobs`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Idempotency-Key': idempotencyKey,
      },
    },
    JSON.stringify({
      upload_id: uploadId,
      operation: 'png_to_webp',
      parameters: { quality: 85 },
    })
  );

  if (jobRes.status !== 202 || !jobRes.body?.job_id) {
    fail(`Job creation failed: ${jobRes.status} ${jobRes.text}`);
  }

  const jobId = jobRes.body.job_id;
  success(`Job queued successfully (Job ID: ${jobId}, Profile: ${jobRes.body.resource_profile})`);

  // Verify Idempotency on duplicate request
  const dupRes = await request(
    `${API_BASE}/v1/jobs`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Idempotency-Key': idempotencyKey,
      },
    },
    JSON.stringify({
      upload_id: uploadId,
      operation: 'png_to_webp',
    })
  );

  if (dupRes.status === 200 && dupRes.body?.id === jobId) {
    success('Idempotent replay verified: returned existing job without duplicate task dispatch');
  } else {
    fail(`Idempotency verification failed: ${dupRes.status} ${dupRes.text}`);
  }

  // 5. Poll Job until Completed
  log('Awaiting worker processing and Output Validator gate...');
  const maxWait = 15;
  const start = Date.now();
  let completed = false;
  let details = null;

  while ((Date.now() - start) / 1000 < maxWait) {
    const pollRes = await request(`${API_BASE}/v1/jobs/${jobId}`);
    if (pollRes.status === 200) {
      details = pollRes.body;
      const status = details.job?.status;
      if (status === 'completed') {
        completed = true;
        break;
      } else if (status === 'failed') {
        fail(`Worker reported job failure: ${pollRes.text}`);
      }
    }
    await new Promise((r) => setTimeout(r, 400));
  }

  if (!completed) {
    fail(`Job timed out after ${maxWait}s. Last state: ${JSON.stringify(details)}`);
  }

  const outputs = details.outputs || [];
  if (outputs.length === 0) {
    fail('Job marked completed but zero outputs registered');
  }

  const out = outputs[0];
  success(`Job completed successfully! Output: '${out.filename}' (${out.file_size} bytes)`);

  if (out.validation_passed) {
    success('Output Validator gate: PASSED (magic bytes, dimensions, non-truncated stream verified)');
  } else {
    fail('Output Validator check failed');
  }

  // 6. Verify Signed Download URL
  const downloadUrl = (out.download_url || '').replace('http://localhost:8085', API_BASE);
  log(`Verifying signed download URL: ${downloadUrl}...`);
  const downloadRes = await request(downloadUrl);
  if (downloadRes.status === 200 && downloadRes.buffer.length > 0) {
    success(`Output downloaded successfully (${downloadRes.buffer.length} bytes)`);
  } else {
    fail(`Download failed with status ${downloadRes.status}`);
  }

  console.log('\n=======================================================');
  console.log('  \x1b[32mALL INTEGRATION & ARCHITECTURE TESTS PASSED (100%)\x1b[0m  ');
  console.log('=======================================================\n');
}

runTests().catch((err) => {
  fail(`Unhandled test error: ${err.message}`);
});

File: tests/test_e2e.py
"""
E2E Integration Test Suite for li-pdf Platform.
Tests:
1. Healthcheck & Capability Registry
2. Presigned Upload Flow
3. Idempotent Job Creation (PNG -> WebP conversion & PDF compression)
4. State Machine (QUEUED -> PROCESSING -> VALIDATING -> COMPLETED)
5. Output Validator Verification & Download
"""

import sys
import time
import json
import urllib.request
import urllib.parse
import os

API_BASE = os.environ.get("API_BASE", "http://localhost:8085")

def log(msg):
    print(f"[\033[94mTEST\033[0m] {msg}")

def success(msg):
    print(f"[\033[92mPASS\033[0m] {msg}")

def fail(msg):
    print(f"[\033[91mFAIL\033[0m] {msg}")
    sys.exit(1)

def request_json(url, method="GET", data=None, headers=None):
    if headers is None:
        headers = {}
    
    req_data = None
    if data is not None:
        req_data = json.dumps(data).encode("utf-8")
        headers["Content-Type"] = "application/json"

    req = urllib.request.Request(url, data=req_data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req) as resp:
            body = resp.read().decode("utf-8")
            if body:
                return resp.status, json.loads(body)
            return resp.status, {}
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8")
        try:
            return e.code, json.loads(body)
        except:
            return e.code, {"error": body}

def run_tests():
    log(f"Connecting to li-pdf API at {API_BASE}...")

    # 1. Healthcheck
    status, res = request_json(f"{API_BASE}/healthz")
    if status == 200 and res.get("status") == "healthy":
        success("Healthcheck passed")
    else:
        fail(f"Healthcheck failed: {status} {res}")

    # 2. Capabilities
    status, res = request_json(f"{API_BASE}/v1/capabilities")
    caps = res.get("capabilities", [])
    if status == 200 and len(caps) > 0:
        success(f"Capability Registry active ({len(caps)} operations registered)")
    else:
        fail(f"Capabilities fetch failed: {status} {res}")

    # 3. Create a synthetic test PNG file
    log("Creating synthetic test PNG...")
    # Minimal 1x1 valid PNG bytes
    png_bytes = b"\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x01\x00\x00\x00\x01\x08\x06\x00\x00\x00\x1f\x15c4\x00\x00\x00\nIDATx\x9cc\x00\x01\x00\x00\x05\x00\x01\r\n-\xb4\x00\x00\x00\x00IEND\xaeB`\x82"
    
    # Presign upload
    status, presign = request_json(f"{API_BASE}/v1/uploads/presign", method="POST", data={
        "filename": "sample_test.png",
        "size_bytes": len(png_bytes),
        "mime_type": "image/png"
    })
    if status != 201 or "upload_url" not in presign:
        fail(f"Presign upload failed: {status} {presign}")
    
    upload_id = presign["upload_id"]
    upload_url = presign["upload_url"]
    success(f"Presigned upload slot created (ID: {upload_id})")

    # Direct Upload to Presigned URL
    log(f"Executing direct upload ({len(png_bytes)} bytes)...")
    req = urllib.request.Request(upload_url, data=png_bytes, headers={"Content-Type": "image/png"}, method="PUT")
    with urllib.request.urlopen(req) as resp:
        if resp.status not in (200, 204):
            fail(f"Direct upload to storage failed with HTTP {resp.status}")
    success("Direct storage payload upload verified")

    # 4. Create Job with Idempotency Key
    idempotency_key = f"e2e_idemp_{int(time.time())}"
    log(f"Creating job: 'png_to_webp' with Idempotency-Key: {idempotency_key}...")
    status, job_res = request_json(f"{API_BASE}/v1/jobs", method="POST", data={
        "upload_id": upload_id,
        "operation": "png_to_webp",
        "parameters": {"quality": 85}
    }, headers={"Idempotency-Key": idempotency_key})

    if status != 202 or "job_id" not in job_res:
        fail(f"Job creation failed: {status} {job_res}")
    
    job_id = job_res["job_id"]
    success(f"Job queued successfully (Job ID: {job_id})")

    # Verify Idempotency on duplicate request
    status, dup_res = request_json(f"{API_BASE}/v1/jobs", method="POST", data={
        "upload_id": upload_id,
        "operation": "png_to_webp",
    }, headers={"Idempotency-Key": idempotency_key})
    if status == 200 and dup_res.get("id") == job_id:
        success("Idempotent replay verified: returned existing job without duplicate creation")
    else:
        fail(f"Idempotency verification failed: {status} {dup_res}")

    # 5. Poll Job until Completed
    log("Awaiting worker processing and Output Validator gate...")
    max_wait = 15
    start = time.time()
    completed = False
    details = {}

    while time.time() - start < max_wait:
        status, details = request_json(f"{API_BASE}/v1/jobs/{job_id}")
        if status == 200:
            j_status = details.get("job", {}).get("status")
            if j_status == "completed":
                completed = True
                break
            elif j_status == "failed":
                fail(f"Worker reported job failure: {details}")
        time.sleep(0.5)

    if not completed:
        fail(f"Job timed out after {max_wait}s. Last state: {details}")

    outputs = details.get("outputs", [])
    if len(outputs) == 0:
        fail("Job marked completed but zero outputs registered")

    out = outputs[0]
    success(f"Job completed successfully! Output: '{out.get('filename')}' ({out.get('file_size')} bytes)")
    
    if out.get("validation_passed"):
        success("Output Validator gate: PASSED (magic bytes, dimensions, non-truncated stream verified)")
    else:
        fail("Output Validator check failed")

    # 6. Verify Signed Download URL
    download_url = out.get("download_url")
    log(f"Verifying signed download URL: {download_url}...")
    with urllib.request.urlopen(download_url) as resp:
        downloaded_bytes = resp.read()
        if len(downloaded_bytes) > 0:
            success(f"Output downloaded successfully ({len(downloaded_bytes)} bytes)")
        else:
            fail("Downloaded file is empty")

    print("\n=======================================================")
    print("  \033[92mALL INTEGRATION & ARCHITECTURE TESTS PASSED (100%)\033[0m  ")
    print("=======================================================\n")

if __name__ == "__main__":
    run_tests()
